### ISSUE TITLE
forgeBSgenomeDataPkgFromNCBI() enhancement: do not download genomic sequences again if already downloaded

### ISSUE THREAD

**Issue by hpages** [Bioconductor/BSgenomeForge#25]
Right now the same FASTA file (`*_genomic.fna.gz` file) gets downloaded again each time we call `forgeBSgenomeDataPkgFromNCBI()` on the same assembly accession. Note that it gets downloaded in the current directory (a.k.a. working directory). More precisely, each new download silently replaces the previous one with the exact same file.

We should download the `*_genomic.fna.gz` file only if it's not already present in the current directory. In order to check for the presence of this file, we need to know its exact name. I just added an internal helper (`get_URL_to_genomic_sequences_from_NCBI()`) that can be used for that. You just need to call `basename()` on the URL returned by `get_URL_to_genomic_sequences_from_NCBI()` to get the name of the file.

`get_URL_to_genomic_sequences_from_NCBI()` is defined in `R/downloadGenomicSequencesFromNCBI.R`. Don't hesitate to take a look there to see how it works and how it's used in `downloadGenomicSequencesFromNCBI()`. (You might need to resync your fork if you.ve not already done that, in order to see the `get_URL_to_genomic_sequences_from_NCBI()` function.)

To summarize, we need to:
- Check whether the `*_genomic.fna.gz` file is present in the current directory. If not then download it. If present then display a message (with `message()`) that says that the file is already in the current directory so will be used.
- This behavior should be documented. A good place for that is the `\details` section which is where we typically provide any additional details about what the function does and how to use it. This is a good place to explain where the FASTA file gets downloaded and what happens if it's already there.

Let me know if you have questions.

Thanks!

**Comment by kakopo** [comment]
Hi Hervé, please take a look at my PR regarding this issue. I hope my approach is alright but corrections are always appreciated, especially regarding the man page edits. Thank you!

**Comment by hpages** [comment]
Looks good.

**R code:**
```
    fasta_file <- basename(get_URL_to_genomic_sequences_from_NCBI
                           (assembly_accession))
```
The line break between the function name and the opening parenthesis makes it hard to read that code. It takes some effort to realize that `(assembly_accession)` is part of a function call. This can easily be avoided by keeping the opening parenthesis on the same line as the function name e.g.:
```
    fasta_file <- basename(get_URL_to_genomic_sequences_from_NCBI(
                                                        assembly_accession))
```
However this is not very pretty either. I think this is a case where doing one function call per line makes the code easier to read:
```
    file_url <- get_URL_to_genomic_sequences_from_NCBI(assembly_accession)
    fasta_file <- basename(file_url)
```
Note that this means that we need to introduce an intermediate variable (`file_url`) and some people don't like to do this. Personally I don't see this as a problem at all, quite the opposite, because this is an opportunity to choose a name for the variable that makes it clear what the intermediate value is. So the whole thing just becomes easier to follow.

FWIW I see a lot of code where the developer tries to pack as much operations as they can per line, e.g. `res <- foo4(foo3(foo2(foo1(x))))`, resulting in code that is very dense and can be hard to read. A popular alternative to this is to use the pipe operator (`|>`) to chain the operations: `res <- x |> foo1() |> foo2() |> foo3() |> foo4()`.
Personally I'm not a big fan of this. I still prefer to break the expression into 4 separated steps, one step per line:
```
    tmp <- foo1(x)
    tmp <- foo2(tmp)
    tmp <- foo3(tmp)
    res <- foo4(tmp)
```

Note that in our case, using the pipe operator would look like this:
```
    fasta_file <- assembly_accession
                  |> get_URL_to_genomic_sequences_from_NCBI()
                  |> basename()
```



**Man page:**

"... to download the compressed FASTA file from NCBI, and stores **them** ...". Shouldn't it be "and stores **it** ..."? (you're referring to the file)

I would even clarify what this FASTA file is about by saying:

"... to download the compressed FASTA file that contains the genomic sequences of the assembly, and stores it in the working directory."

Also I would say, "It typically makes use of **the** \code{downloadGenomicSequencesFromNCBI} **utility function** to download ..."

I would try to keep this part shorter/simpler:
```
However, should the file
already exist in the working directory, it is not re-downloaded. Instead, it
makes use of \code{get_URL_to_genomic_sequences_from_NCBI} to obtain the file
name and ensure it exists in the working directory. Hereafter, the rest of the
function is executed.
```
The user doesn't know about the `get_URL_to_genomic_sequences_from_NCBI` function and they don't have access to it (the function is not exported and not documented) so providing them with this level of detail is not really necessary or useful to them.

So I would just say something like:
```
However, if the file already exists in the working directory, then it is used and not downloaded again.
```
That's all! I don't think you need the last sentence ("Hereafter, the rest of the function is executed.")

**Comment by hpages** [comment]
I forgot to mention that you don't need to use `paste0()` here:
```
    if (file.exists(fasta_file)) {
        message(wmsg(paste0("The file ", fasta_file, " is already in the current
                       directory so it will be used.")))
    } else {
        fasta_file <- downloadGenomicSequencesFromNCBI(assembly_accession)
    }
```
because `wmsg()` already takes care of that.

Also using a string literal that spans more than one line should be avoided. It's better to break the long string in smaller strings (`wmsg()` will paste them together), and separate them with commas:
```
        message(wmsg("The file ", fasta_file, " is already in the current ",
                     "directory so will be used."))
```
(note that I also got rid of "it" in "so will be used")

**Comment by kakopo** [comment]
Thank you for the feedback. I agree that writing 
```
fasta_file <- basename(get_URL_to_genomic_sequences_from_NCBI
                           (assembly_accession))
```
was not the best looking approach. I've improved on the code with the suggestion you shared. I've also updated the man page. Please take a look. Thanks!

**Comment by hpages** [comment]
Alright, PR #28 is merged. Please resync your fork.

I'm going to try to add some guidance and tips in issue #24 for how to procede with the vignette.
